src/agents/coordinator.ts

```diff
@@ -80,6 +80,15 @@ export class DkgCoordinatorAgent {
     return Coordinator.rituals(ritualId);
   }
 
+  public static async getTimeout(
+    provider: ethers.providers.Web3Provider
+): Promise<number> {
+    const Coordinator = await this.connectReadOnly(provider);
+    const timeout = await Coordinator.timeout();
```

Comment:
Timeout is the number of seconds counted since the ritual started, right? What happens if my transaction gets stuck (low gas etc.) but I'm still counting from when I send the transaction? Should we instead count the timeout from the "ritual start" event or some other cut-off?

Reply:
No, `timeout` is a variable on the Coordinator https://github.com/nucypher/nucypher-contracts/blob/da35b9d9f13ddebd7ce7bbbc3fbc7d5fb0d95411/contracts/contracts/coordination/Coordinator.sol#L61

It's set at Coordinator level and applies to all rituals

Reply:
I think I'm missing something here. Timeout is defined as [`ritual.initTimestamp + timeout < block.timestamp`](https://github.com/nucypher/nucypher-contracts/blob/da35b9d9f13ddebd7ce7bbbc3fbc7d5fb0d95411/contracts/contracts/coordination/Coordinator.sol#L76-L77), and the initial value of `ritual.initTimestamp` is taken from [`block.timestamp`](https://github.com/nucypher/nucypher-contracts/blob/da35b9d9f13ddebd7ce7bbbc3fbc7d5fb0d95411/contracts/contracts/coordination/Coordinator.sol#L127). So it seems to me like the `timeout` denotes some number of seconds, and the timeout occurs after we reach some number of seconds after the ritual started.

Do you see any edge cases here? If we use `setTimeout`, is it going to match exactly the calculation performed in ritual state checks? I.e. `ritual.initTimestamp + timeout < block.timestamp`

Reply:
>`timeout` denotes some number of seconds, and the timeout occurs after we reach some number of seconds after the ritual started.
Yes, agree

`setTimeout` is changing the config of Coordinator, so not sure that it helps us.

But i see what your original comment implied now. When we do:
```ts
const ritualId = await DkgCoordinatorAgent.initializeRitual(
      web3Provider,
      ursulas.sort()
    );
```
The fact that we `await` doesn't always mean that the transaction went through? and it could therefore be sitting around in the mempool whilst the `Promise.race` starts counting down. Is that correct?

Reply:
`initializeRitual` is blocked until the ritual is started (and some event is emitted, etc.) so it may not be an issue after all.

I was trying to give a concrete example of something that @manumonti already hinted on in his comment:
> Thinking aloud, since the "timer" that runs in nucypher-ts and the "timer" than runs in Coordinator contract are not the same and probably, due to the network delays, one of them will reach the deadline before the other... I wonder if the difference between these two deadlines is significant or not

I didn't make a full analysis on what are the different edge cases we may have here, but one other that comes to mind is where we're awaiting for `timeout` and the `Coordinator` admin changes the value of `timeout`. I think this case and some basket of other cases can be handled by replacing `setTimeout` by `setInterval` and by redoing the contract calculation `ritual.initTimestamp + timeout < block.timestamp` in JS. Or alternatively, querying the contract for the ritual state. The latter seems to be more robust.